File: backend/agent/nodes/human_review.py
from typing import Dict, Any
from langgraph.types import interrupt

def human_review_node(state: Dict[str, Any]) -> Dict[str, Any]:
    # Pauses graph execution and sends details to the client
    # The API endpoint will catch the interrupt and return a pending approval state
    decision = interrupt({
        "question": f"This query will {state.get('mutation_type')} data. Do you want to proceed?",
        "sql": state.get("generated_sql"),
        "mutationType": state.get("mutation_type")
    })

    if decision == 'reject':
        return {
            "human_approved": False,
            "final_response": {
                "type": "answer",
                "message": "Query execution cancelled by user."
            }
        }

    return {"human_approved": True}

File: lib/agent/nodes/human-review.ts
// lib/agent/nodes/human-review.ts
// Uses LangGraph interrupt() to pause for user approval on mutation queries.
import { interrupt } from "@langchain/langgraph";
import type { AgentStateType } from "../state";

export function humanReviewNode(
  state: AgentStateType,
): Partial<AgentStateType> {
  // This call pauses graph execution and sends state to the client.
  // The API route catches the interrupt and returns pendingApproval: true.
  const decision = interrupt({
    question: `This query will ${state.mutationType} data. Do you want to proceed?`,
    sql: state.generatedSQL,
    mutationType: state.mutationType,
  });

  // When resumed, decision is either "approve" or "reject"
  if (decision === "reject") {
    return {
      humanApproved: false,
      finalResponse: {
        type: "answer",
        message: "Query execution cancelled by user.",
      },
    };
  }

  return { humanApproved: true };
}
